### ISSUE TITLE
Issues with accuracy and loss value fluctuations in each epoch, and the setting of random number seeds.

### ISSUE THREAD

**Issue by yang123-xd** [NVlabs/MambaVision#82]
Thank you for your great ideas and word.

I am working on a person re-identification task and have encountered an issue. I have a complete set of code that includes data loading, model definition, loss calculation, test accuracy computation, etc., and it can normally perform the person re-identification task with consistent training accuracy and loss values in each epoch without random fluctuations.

Now, I have replaced the backbone model from the original ResNet with the MambaVision model, and the training accuracy and loss values fluctuate in each epoch. I have set global random number seeds and suspect that the randomness originates from the MambaVision code or the timm library. Has the author encountered similar issues with loss fluctuations? (Since the rest of the code remains unchanged, I suspect the randomness is from the MambaVision code or the timm library.)

**Comment by ahatamiz** [comment]
Thanks for reaching out @yang123-xd ! 

Have you considered using a smaller learning rate ? the random fluctuations in this case could indeed indicate that training dynamics is sub-optimal. 

We did not encounter such fluctuations in training of the base model (ImageNet) or downstream tasks. And I also don't think the issue has to do with random seeds. Though various random seeds may show slightly different results, and training dynamics, it must not significantly change the outcome.

**Comment by yang123-xd** [comment]
> Thanks for reaching out [@yang123-xd](https://github.com/yang123-xd) !
> 
> Have you considered using a smaller learning rate ? the random fluctuations in this case could indeed indicate that training dynamics is sub-optimal.
> 
> We did not encounter such fluctuations in training of the base model (ImageNet) or downstream tasks. And I also don't think the issue has to do with random seeds. Though various random seeds may show slightly different results, and training dynamics, it must not significantly change the outcome.










    Thank you for your prompt response.
    My learning rate is set to 0.00035, and the weight decay coefficient is 5e-4. I have tried adjusting the learning rate, but regardless of the value, the results fluctuate in each training run.
    My code is for a person re-identification task, and I have set a global random seed at the beginning of training as follows:

![Image](https://github.com/user-attachments/assets/1f54004f-a9c1-4389-9b98-5a2f13ec897c)

Previously, when using a ResNet backbone, the training results were consistent across all runs.

![Image](https://github.com/user-attachments/assets/350e3e4d-be08-43c3-a47e-5cb36a49f4fe)

I am very interested in your research and wanted to try the MambaVision model in this field. However, after replacing the backbone network from ResNet with MambaVision (with only one line of code changed), the training results now fluctuate.

![Image](https://github.com/user-attachments/assets/b952458c-77f1-4124-af9d-65b85bbf607d)

I have reviewed the source code of MambaVision and found it includes modules built on PyTorch, which should theoretically be affected by the global random seed to stabilize accuracy. However, the code also imports modules from the timm and mamba_ssm libraries. I suspect these imports are the root cause of the issue, as their randomness might not be covered by the global seed, leading to fluctuations.

This is similar to a previous problem I encountered with the PYG library: using its GCN modules caused accuracy fluctuations, but the issue was resolved after implementing a custom GCN from scratch (for reference: https://blog.csdn.net/lxzlxzlxzzads/article/details/135095393).

Currently, I have tried reimplementing all functions imported from timm and mamba_ssm manually, but the problem persists. Do you have any suggestions for resolving this or troubleshooting steps to identify the root cause?


result1

![Image](https://github.com/user-attachments/assets/15df8924-4f4b-479d-9def-e5ac2bc44692)
result2

![Image](https://github.com/user-attachments/assets/4bc7c18d-72ef-448d-9e8d-9681b04f9894)
result3

![Image](https://github.com/user-attachments/assets/bc307b7c-8abe-42bb-a419-21e32621a03a)

**Comment by ahatamiz** [comment]
Thanks @yang123-xd for raising the issue. While it is hard for me to exactly pin-point the root cause, here's a couple of things that comes to my mind:

(1) Even though you already set a global random seed, it is still worth ensuring you have all relevant seeds set, and that you configure PyTorch’s additional flags for deterministic behavior where applicable. For instance:

```
import random
import numpy as np
import torch

seed = 42
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)  # if you use multiple GPUs

torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
torch.use_deterministic_algorithms(True)
```

Also re-check if any data augmentation or random transformation is not fixed. For instance, timm might apply certain default random augmentations or mixup/cutmix if used in classification scenarios.


(2) try setting drop_path_rate in mambavision to zero to eliminate any randomness from the backbone. This is just a sanity check to ensure the fluctuations don't arise from mambavision randomness. 

(3) I briefly talked about this in the previous response, but you should verify if you fluctuations are happening "in-epoch" or from "run-to-run". Basically, if you observe that, within a single training epoch, your losses or accuracy bounce around unpredictably between iterations, that sometimes indicates a more fundamental training stability issue. This can be a symptom of a too-high learning rate, exploding gradients. But, if your first run gets, say, final mAP of 80%, the second run gets 75%, the third 78%, etc. (despite using the same seed), that typically means some hidden randomness or non-deterministic GPU ops.

(4) Try a bigger variant of MambaVision. Maybe use MambaVision-L2 since it is a more powerful model and see if you still get these fluctuations. 

(5) Temporarily disable or freeze MambaVision backbone to see if the source of fluctuation comes from the backbone. 

(6) Here's just a bunch of suggestions to further stabilize the training:

- Lower learning rate (as you already tried).
- Gradual warm-up.
- More stable optimizers (e.g., use AdamW vs. SGD).
- Different weight decays.
- Turn off AMP and use float32 if possible 

Hope this helps !

**Comment by ahatamiz** [comment]
Unfortunately, I don't have more information to provide beyond what I presented in my response.

**Comment by koro-github** [comment]
> Unfortunately, I don't have more information to provide beyond what I presented in my response.


Thank you for your response. I will continue to try to solve this problem and sincerely appreciate your help.

**Comment by ahatamiz** [comment]
We did not observe inconsistent results for downstream tasks such as object detection. In fact, we just released the code and models for object detection (please see [here](https://github.com/NVlabs/MambaVision/tree/main/object_detection)).  

Hope this is useful.
